pintracker/stateless/stateless.go

```diff
@@ -0,0 +1,490 @@
+		"",
+		"Cluster",
+		"IPFSPinLsCid",
+		api.PinCid(c).ToSerial(),
+		&ips,
+	)
+	if err != nil {
+		logger.Error(err)
+		return api.PinInfo{}
+	}
+
+	pi := api.PinInfo{
+		Cid:    c,
+		Peer:   spt.peerID,
+		Status: ips.ToTrackerStatus(),
+	}
+
+	return pi
+}
+
+// SyncAll verifies that the statuses of all tracked Cids (from the shared state)
+// match the one reported by the IPFS daemon. If not, they will be transitioned
+// to PinError or UnpinError.
+//
+// SyncAll returns the list of local status for all tracked Cids which
+// were updated or have errors. Cids in error states can be recovered
+// with Recover().
+// An error is returned if we are unable to contact the IPFS daemon.
+func (spt *Tracker) SyncAll() ([]api.PinInfo, error) {
+	// get ipfs status for all
+	localpis, err := spt.localStatus(false)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	for _, p := range spt.optracker.FilterOps(optracker.OperationPin, optracker.PhaseError) {
+		if _, ok := localpis[p.Cid().String()]; ok {
+			spt.optracker.Clean(p)
+		}
+	}
+
+	for _, p := range spt.optracker.FilterOps(optracker.OperationUnpin, optracker.PhaseError) {
+		if _, ok := localpis[p.Cid().String()]; ok {
+			spt.optracker.Clean(p)
+		}
+	}
+
+	return spt.getErrorsAll(), nil
+}
+
+// Sync returns the updated local status for the given Cid.
+func (spt *Tracker) Sync(c *cid.Cid) (api.PinInfo, error) {
+	return spt.Status(c), nil
+}
+
+// RecoverAll attempts to recover all items tracked by this peer.
+func (spt *Tracker) RecoverAll() ([]api.PinInfo, error) {
+	statuses := spt.StatusAll()
+	resp := make([]api.PinInfo, 0)
+	for _, st := range statuses {
+		r, err := spt.Recover(st.Cid)
+		if err != nil {
+			return resp, err
+		}
+		resp = append(resp, r)
+	}
+	return resp, nil
+}
+
+// Recover will re-track or re-untrack a Cid in error state,
+// possibly retriggering an IPFS pinning operation and returning
+// only when it is done.
+func (spt *Tracker) Recover(c *cid.Cid) (api.PinInfo, error) {
+	logger.Infof("Attempting to recover %s", c)
+	pInfo, ok := spt.getError(c)
+	if !ok {
+		return spt.Status(c), nil
+	}
+
+	var err error
+	switch pInfo.Status {
+	case api.TrackerStatusPinError:
+		err = spt.enqueue(api.PinCid(c), optracker.OperationPin)
+	case api.TrackerStatusUnpinError:
+		err = spt.enqueue(api.PinCid(c), optracker.OperationUnpin)
+	}
+	if err != nil {
+		return spt.Status(c), err
+	}
+
+	pi, ok := spt.optracker.GetExists(c)
+	if !ok {
+		return spt.Status(c), nil
+	}
+
+	return pi, nil
+}
+
+func (spt *Tracker) ipfsStatusAll() (map[string]api.PinInfo, error) {
+	var ipsMap map[string]api.IPFSPinStatus
+	err := spt.rpcClient.Call(
+		"",
+		"Cluster",
+		"IPFSPinLs",
+		"recursive",
+		&ipsMap,
+	)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+	pins := make(map[string]api.PinInfo, 0)
+	for cidstr, ips := range ipsMap {
+		c, err := cid.Decode(cidstr)
+		if err != nil {
+			logger.Error(err)
+			continue
+		}
+		p := api.PinInfo{
```

Comment:
Set `.TS` to `time.Now()`, as it reflects the last time the record was updated.

Reply:
But this is just retrieving the record? It isn't actually updating the record.

Reply:
Read it as: "As of <timestamp>, the state of this Cid was <status>".

Reply:
This is still missing

Reply:
✅